cura/Machines/Models/MachineListModel.py

```diff
@@ -0,0 +1,96 @@
+    MachineCountRole = Qt.ItemDataRole.UserRole + 7
+
+    def __init__(self, parent=None) -> None:
+        super().__init__(parent)
+
+        self._catalog = i18nCatalog("cura")
+
+        self.addRoleName(self.NameRole, "name")
+        self.addRoleName(self.IdRole, "id")
+        self.addRoleName(self.HasRemoteConnectionRole, "hasRemoteConnection")
+        self.addRoleName(self.MetaDataRole, "metadata")
+        self.addRoleName(self.IsOnlineRole, "isOnline")
+        self.addRoleName(self.MachineTypeRole, "machineType")
+        self.addRoleName(self.MachineCountRole, "machineCount")
+
+        self._change_timer = QTimer()
+        self._change_timer.setInterval(200)
+        self._change_timer.setSingleShot(True)
+        self._change_timer.timeout.connect(self._update)
+
+        # Listen to changes
+        CuraContainerRegistry.getInstance().containerAdded.connect(self._onContainerChanged)
+        CuraContainerRegistry.getInstance().containerMetaDataChanged.connect(self._onContainerChanged)
+        CuraContainerRegistry.getInstance().containerRemoved.connect(self._onContainerChanged)
+        self._updateDelayed()
+
+    def _onContainerChanged(self, container) -> None:
+        """Handler for container added/removed events from registry"""
+
+        # We only need to update when the added / removed container GlobalStack
+        if isinstance(container, GlobalStack):
+            self._updateDelayed()
+
+    def _updateDelayed(self) -> None:
+        self._change_timer.start()
+
+    def _update(self) -> None:
+        self.setItems([])  # Clear items
+
+        other_machine_stacks = CuraContainerRegistry.getInstance().findContainerStacks(type="machine")
+
+        abstract_machine_stacks = CuraContainerRegistry.getInstance().findContainerStacks(type = "abstract_machine")
+        abstract_machine_stacks.sort(key = lambda machine: machine.getName(), reverse = True)
+
+        for abstract_machine in abstract_machine_stacks:
+            online_machine_stacks = AbstractMachine.getMachines(abstract_machine, online_only = True)
+
+            # Create a list item for abstract machine
+            self.addItem(abstract_machine, len(online_machine_stacks))
+
+            # Create list of machines that are children of the abstract machine
+            for stack in online_machine_stacks:
+                self.addItem(stack)
+                # Remove this machine from the other stack list
+                other_machine_stacks.remove(stack)
+
+        for stack in other_machine_stacks:
+            self.addItem(stack)
+
+    def addItem(self, container_stack: ContainerStack, machine_count: int = 0) -> None:
+        if parseBool(container_stack.getMetaDataEntry("hidden", False)):
+            return
+
+        is_online = parseBool(container_stack.getMetaDataEntry("is_online", False))
+        if container_stack.getMetaDataEntry("type") == "abstract_machine":
+            is_online = True
```

Comment:
Why don't we add "is_online" as a property to the meta data of abstract machine? currently we only add cloud printers to the abstract machines so that would always be true.

It is a bit annoying if we manually have to add the abstract machines to online printers in all instances we implement something similar